config-templates/pgsql_tables.sql

```diff
@@ -0,0 +1,40 @@
+--Statistics for IdPs
+CREATE TABLE statistics (
+    year bigint NOT NULL,
+    month bigint NOT NULL,
+    day bigint NOT NULL,
+    sourceidp character varying(255) NOT NULL,
+    service character varying(255) NOT NULL,
+    count bigint,
+    PRIMARY KEY (year, month, day, sourceIdp, service)
+);
+
+CREATE INDEX statistics_i1 ON statistics (sourceIdp);
+CREATE INDEX statistics_i2 ON statistics (service);
+
+CREATE TABLE statistics_detail (
+    year bigint NOT NULL,
+    month bigint NOT NULL,
+    day bigint NOT NULL,
+    sourceidp character varying(255) NOT NULL,
+    service character varying(255) NOT NULL,
+    userid character varying(255) NOT NULL,
+    count bigint,
+    PRIMARY KEY (year, month, day, sourceIdp, service, userid)
+);
+
+CREATE INDEX statistics_detail_i1 ON statistics (sourceIdp);
+CREATE INDEX statistics_detail_i2 ON statistics (service);
```

Comment:
Do we need to explicitly create an INDEX for columns which are part of the table's primary key? If yes, then `statistics` needs to be changed to `statistics_detail`

Reply:
The tablename has been fixed

Reply:
The primary key constraint automatically creates a multi-column B-tree index:

```
Table "public.statistics_detail"
  Column   |          Type          | Modifiers | Storage  | Stats target | Description 
-----------+------------------------+-----------+----------+--------------+-------------
 year      | bigint                 | not null  | plain    |              | 
 month     | bigint                 | not null  | plain    |              | 
 day       | bigint                 | not null  | plain    |              | 
 sourceidp | character varying(255) | not null  | extended |              | 
 service   | character varying(255) | not null  | extended |              | 
 userid    | character varying(255) | not null  | extended |              | 
 count     | bigint                 |           | plain    |              | 
Indexes:
    "statistics_detail_pkey" PRIMARY KEY, btree (year, month, day, sourceidp, service, userid)
```

Based on the [documentation](https://www.postgresql.org/docs/9.6/indexes-multicolumn.html):

> A multicolumn B-tree index can be used with query conditions that involve any subset of the index's columns, but the index is most efficient when there are constraints on the leading (leftmost) columns.

> This index could in principle be used for queries that have constraints on b and/or c with no constraint on a — but the entire index would have to be scanned, so in most cases the planner would prefer a sequential table scan over using the index.

> Multicolumn indexes should be used sparingly. In most situations, an index on a single column is sufficient and saves space and time. Indexes with more than three columns are unlikely to be helpful unless the usage of the table is extremely stylized.

So we most probably need indexes on individual columns or specific combinations of columns frequently used in queries. Investigate if we need:
1. btree index on `(userid)`
1. btree index on `(year, userid)`
2. btree index on `(year, month, userid)`
3. btree index on `(year, month, day, userid) `

\cc @lionick